common/src/main/java/com/wynntils/mc/mixin/GuiGraphicsMixin.java

```diff
@@ -0,0 +1,118 @@
+import java.util.List;
+import java.util.Optional;
+import net.minecraft.client.gui.Font;
+import net.minecraft.client.gui.GuiGraphics;
+import net.minecraft.client.gui.screens.Screen;
+import net.minecraft.network.chat.Component;
+import net.minecraft.world.inventory.tooltip.TooltipComponent;
+import net.minecraft.world.item.ItemStack;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.ModifyVariable;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
+
+@Mixin(GuiGraphics.class)
+public abstract class GuiGraphicsMixin {
+    // Note: Call site 1 of 3 of ItemTooltipRenderEvent. Check the event class for more info.
+    @WrapOperation(
+            method = "renderTooltip(Lnet/minecraft/client/gui/Font;Lnet/minecraft/world/item/ItemStack;II)V",
+            at =
+                    @At(
+                            value = "INVOKE",
+                            target =
+                                    "Lnet/minecraft/client/gui/GuiGraphics;renderTooltip(Lnet/minecraft/client/gui/Font;Ljava/util/List;Ljava/util/Optional;II)V"))
+    private void renderTooltipPre(
+            GuiGraphics instance,
+            Font font,
+            List<Component> tooltipLines,
+            Optional<TooltipComponent> visualTooltipComponent,
+            int mouseX,
+            int mouseY,
+            Operation<Void> operation,
+            @Local ItemStack itemStack) {
+        ItemTooltipRenderEvent.Pre event = new ItemTooltipRenderEvent.Pre(
+                (GuiGraphics) (Object) this,
+                itemStack,
+                Screen.getTooltipFromItem(McUtils.mc(), itemStack),
+                mouseX,
+                mouseY);
+        MixinHelper.post(event);
+
+        if (event.isCanceled()) return;
+
+        operation.call(instance, font, event.getTooltips(), event.getItemStack().getTooltipImage(), mouseX, mouseY);
+    }
+
+    @Inject(
+            method = "renderTooltip(Lnet/minecraft/client/gui/Font;Lnet/minecraft/world/item/ItemStack;II)V",
+            at = @At("RETURN"))
+    private void renderTooltipPost(Font font, ItemStack itemStack, int mouseX, int mouseY, CallbackInfo ci) {
+        MixinHelper.post(new ItemTooltipRenderEvent.Post((GuiGraphics) (Object) this, itemStack, mouseX, mouseY));
+    }
+
+    @ModifyVariable(
+            method =
+                    "renderItemDecorations(Lnet/minecraft/client/gui/Font;Lnet/minecraft/world/item/ItemStack;IILjava/lang/String;)V",
+            at = @At("HEAD"),
+            ordinal = 0,
+            argsOnly = true)
+    private String renderItemDecorations(
+            String text,
+            Font font,
+            ItemStack itemStack,
+            int x,
+            int y,
+            String ignored,
+            @Share("wynntilsCountOverlayColor") LocalIntRef wynntilsCountOverlayColor) {
```

Comment:
I have not seen this before. Is it from the mixin extension package? Does it create a "hidden" field which can be accessed only from those two methods?

Reply:
https://github.com/LlamaLad7/MixinExtras/wiki/Local#mutability

It's a neat thing. I tried using it with Inject-cancel to simplify code instead of WrapOperation, but it did not work. I have not submitted an issue yet; I am not sure if it's intended to change method arguments as well. But from the examples it feels like it.